text/0000-with-bounds-in-pi-types.md

```diff
@@ -0,0 +1,685 @@
+- Feature Name: with-bounds-in-pi-types
+- Start Date: 2017-02-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+This RFC is a part of the [pi type
+trilogy](https://github.com/rust-lang/rfcs/issues/1930).
+
+This RFC introduces `with` bounds on type level values, making it possible to
+set constraints on the constant parameters introduced in implementations,
+functions, structs, and other items.
+
+# Motivation
+[motivation]: #motivation
+
+There are many usecases. These allows certain often requested features to live
+in standalone libraries (e.g.,
+[bounded-integers](#bounded-integersinterval-arithmetics), [type level
+numerals](#array-generics), [statically checked
+indexing](#statically-checked-indexing), lattice types).
+
+I've listed examples of such usecases and their implementations below.
+
+## Examples
+
+### Bounded integers/interval arithmetics
+
+One can define the so called "bounded integers" (integers which carry an
+upper and lower bound, checked at compile time):
+
+```rust
+use std::ops;
+
+/// A bounded integer.
+///
+/// This has two value parameter, respectively representing an upper and a lower bound.
+pub struct BoundedInt<const lower: usize, const upper: usize> {
```

Comment:
Is there a missing `with lower <= upper` here? The rest of the code seemed to assume it.